packages/fiber/src/core/renderer.ts

```diff
@@ -46,87 +49,83 @@ interface Catalogue {
-  [THREE.Sphere, [new THREE.Vector3(), -1]],
-  [THREE.Ray, [new THREE.Vector3(), new THREE.Vector3(0, 0, -1)]],
-  [THREE.Spherical, [1, 0, 0]],
-  [THREE.Triangle, [new THREE.Vector3(), new THREE.Vector3(), new THREE.Vector3()]],
-].forEach(([type, args]) => defaultMap.set(type, args))
-
 let emptyObject = {}
 let catalogue: Catalogue = {}
 let extend = (objects: object): void => void (catalogue = { ...catalogue, ...objects })
 
-function createRenderer<TCanvas>(roots: Map<TCanvas, Root>, invalidate: (state?: RootState) => void) {
+const filterProps = ['children', 'key', 'ref']
+
+function createRenderer<TCanvas, TRoot = Root>(roots: Map<TCanvas, TRoot>) {
   function applyProps(instance: Instance, newProps: InstanceProps, oldProps: InstanceProps = {}, accumulative = false) {
     // Filter equals, events and reserved props
-    const localState = (instance?.__r3f ?? {}) as LocalState
+    const localState = (instance.__r3f ?? {}) as LocalState
     const root = localState.root
     const rootState = root?.getState() ?? {}
     const sameProps: string[] = []
     const handlers: string[] = []
 
-    let i
-    let keys = Object.keys(newProps)
+    const newMemoizedProps: { [key: string]: any } = {}
+
+    Object.entries(newProps).forEach(([key, entry]) => {
+      // we don't want children, ref or key in the memoized props
+      if (filterProps.indexOf(key) === -1) {
+        newMemoizedProps[key] = entry
+      }
+    })
+
+    if (localState.memoizedProps.args) {
+      newMemoizedProps.args = localState.memoizedProps.args
+    }
+
+    instance.__r3f.memoizedProps = newMemoizedProps
 
-    for (i = 0; i < keys.length; i++) {
-      if (is.equ(newProps[keys[i]], oldProps[keys[i]])) sameProps.push(keys[i])
+    for (let newPropsKey in newProps) {
+      if (is.equ(newProps[newPropsKey], oldProps[newPropsKey])) {
+        sameProps.push(newPropsKey)
+      }
 
       // Event-handlers ...
       //   are functions, that
       //   start with "on", and
       //   contain the name "Pointer", "Click", "ContextMenu", or "Wheel"
-      if (is.fun(newProps[keys[i]]) && keys[i].startsWith('on')) {
+      if (is.fun(newProps[newPropsKey]) && newPropsKey.startsWith('on')) {
         if (
-          keys[i].includes('Pointer') ||
-          keys[i].includes('Click') ||
-          keys[i].includes('ContextMenu') ||
-          keys[i].includes('Wheel')
+          newPropsKey.includes('Pointer') ||
+          newPropsKey.includes('Click') ||
+          newPropsKey.includes('ContextMenu') ||
+          newPropsKey.includes('Wheel')
         ) {
-          handlers.push(keys[i])
+          handlers.push(newPropsKey)
         }
       }
     }
 
     // Catch props that existed, but now exist no more ...
     const leftOvers = [] as string[]
-    keys = Object.keys(oldProps)
     if (accumulative) {
-      for (i = 0; i < keys.length; i++) if (!newProps.hasOwnProperty(keys[i])) leftOvers.push(keys[i])
+      for (let oldPropKey in oldProps) {
```

Comment:
These were turned from .foreach into for some time ago to speed up the hot path, is for in comparable?

Reply:
Ah no, I think it's probably lower, I can revert this change.

Reply:
We lose out on a couple minor optimizations. For example, the `let oldPropKey` is being defined each iteration whereas we were reusing `i` before. /shrug

Reply:
Yeah, revert on these is better, unless we find something faster 🤷